nameserver/zone.go

```diff
@@ -121,6 +122,37 @@ func (zone *ZoneDb) AddRecord(ident string, name string, ip net.IP) error {
 	return nil
 }
 
+func (zone *ZoneDb) DeleteRecords(ident *string, ip *net.IP, name *string) (count int) {
+	zone.mx.Lock()
+	defer zone.mx.Unlock()
+
+	var fqdn *string
+	if name != nil {
+		temp := dns.Fqdn(*name)
+		fqdn = &temp
+	}
+
+	match := func(r dbRecord) bool {
+		return (ident == nil || r.Ident == *ident) &&
+			(ip == nil || r.IP.Equal(*ip)) &&
+			(fqdn == nil || r.Name == *fqdn)
+	}
+
+	w := 0 // write index
+
+	for _, r := range zone.recs {
+		if !match(r) {
+			zone.recs[w] = r
+			w++
+		}
+	}
+
+	count = len(zone.recs) - w
+	zone.recs = zone.recs[:w]
+
+	return
+}
+
```

Comment:
Good point. @rade also points out that it forces a traversal of the underlying data structure, which would prevent optimisation.

In terms of other use cases - remove all records (*, *, *) perhaps? That being said, actively *preventing* combinations which we don't consider to be useful involves a bunch of extra checks - do we want to do that and return an error?